[WIP] Add priority and fairness (#15899)
kubernetes/website: content/en/docs/concepts/cluster-administration/priority-and-fairness.md

---

> title: Priority and Fairness in Handling Requests

Reviewer: Suggested title:
```suggestion
title: API request priority
```

Author: done

content_template: templates/concept
weight: 20
---

> 

Reviewer 2: After a quick read, I am wondering if this page belongs under `docs/reference/using-api/api-concepts`, or near the other API content pages,
https://5d579d7acb65dd0009e4cc90--kubernetes-io-vnext-staging.netlify.com/docs/reference/using-api/api-concepts/

Author: Frankly, I am not impressed with the fact that "using the Kubernetes API" and "accessing the API" are two different sections; it is not entirely obvious to me what is the rule the defines what goes in which section.  But the page being reviewed here is more like the authentication and authorization stuff in "accessing the API" (being about configured control over requests) than it is like the stuff in the "using" section.

Reviewer: It feels to me like there's room for a short new page about this page in the reference section as well.
(one approach: get this merged to `dev-1.16` and then log an issue about adding the reference page).

{{% capture overview %}}

{{< feature-state state="alpha" >}}

> The control plane of Kubernetes is itself a resource (broadly

Reviewer 2: hello @MikeSpreitzer . A few initial comments:
- It may be helpful to change the headings, `Outline` and `Status`, to something more meaningful.
This first section could use the heading, `Overview`.
- This feature is called `RequestManagement`? You could use this name to create the title, such as `API Server Request Management`, or something similar.
- There are other docs pages that describe how to enable a feature. Perhaps use similar language to describe enabling the feature. One example: https://kubernetes.io/docs/reference/using-api/api-concepts/#enable-the-server-side-apply-alpha-feature 
- It looks like the feature is already listed in the feature gates table. Check that the version and definition summary are correct? https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
- Could this page link to definitions for `WATCHes` and `EXECUTes`?

Author: Yes, this feature is called `RequestManagement` but that was a compromise name adopted in early discussions before we really understood this work well.  I think "API priority and fairness" is a better description; "request management" is very vague and could include the functionality of this feature and/or other functionality.  So I would not like the poor choice of feature name to influence the doc any more than necessary.  In fact, I might even suggest changing the feature name.

Reviewer: `kubernetes EXEC` (in the API) sense seems a hard thing to find via web search. I don't think there's a page about it in the Kubernetes docs. If there's a page anywhere on the web I can try to read that page, understand the contnts, and add something about that to the Kubernetes docs.

(I don't know much about Kubernetes at that level of API detail).

Reviewer: It's definitely a good idea to use a feature-state shortcode and to hyperlink to https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

Maybe in the future there'll be more automation so that the actual feature status from [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes/) gets copied to the relevant parts of the website. Right now that's not in place.
The website has had, and still has, pages that describe the status of a feature the last time that a particular page was touched. At the time of writing [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#limitations) was a good example of this, talking about versions back to v1.5 (!)

With that in mind, I favor being a little DRY and leaving the status information for the feature gates page.

> construed) to which access is managed.  In particular, each apiserver

Reviewer: I recommend using [glossary shortcodes](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#glossary) for “API server” and possibly also “control plane”.

Author: I do not see entries for those in the glossary; are you suggesting I add them?
(There _is_ an entry for "kube apiserver" but there are other apiservers and this feature applies to those too.)

Reviewer: That highlights a gap in the docs, so I logged issue #15957.

applies some prioritization and fairness in its handling of requests.
This includes not only kube-apiservers but aggregated apiservers too.
Prioritization and fairness applies only to requests that are not
_long running_.  Long running requests are WATCHes and EXECUTEs.

{{% /capture %}}

{{% capture body %}}

> ## Status

Reviewer: I think for a concept page it's appropriate to outline both the existing and proposed behavior.
(if that sounds like a good idea, I'll offer to help with documenting the current behavior, and that could potentially land in `master` even before the 1.16 release).

Author: Are you suggesting one page that documents both, or one page that documents the old and one that documents the new?

Reviewer: Either's fine by me.

This feature is named "RequestManagement" and is alpha in
release 1.16.  As an alpha feature, it is disabled by default.  When
enabled it replaces a request filter that enforces simple aggregate
concurrency limits with this feature, which can be seen as a more
general version of that earlier functionality.

> ## Outline

Reviewer: This text might belong in the overview section.

Reviewer 2: 👍

Author: The outline has to be a single paragraph, right?

Reviewer 2: Not sure what you mean. Is this page supposed to be a blog or content page?

Author: This is a page for the CONCEPTS section of the documentation, and I thought I read somewhere that the overview of a CONCEPTS page has to be a single paragraph.

Reviewer: It ought to be short, but if you need to use 2 paragraphs then I think that can be justified. A 2 paragraph overview seems better to me than 2 separate overview-like sections.

> Each apiserver independently applies prioritization and fariness,

Reviewer: nit:
```suggestion
Each apiserver independently applies prioritization and fairness,
```

Alternatively:
```suggestion
If the `RequestManagement` [feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/] is enabled, the API server prioritizes incoming requests, based on the
configuration you have set.
```
(fine to wrap over several lines, of course)
```

Reviewer: Depending on other changes, here might be a good point to introduce the resource names:
`PriorityLevelConfiguration` and `FlowSchema`.

Author: I thought about that, and deliberately draw a distinction between (a) the `PriorityLevelConfiguration` Kind of API object and (b) the concept of a _priority level_.  A priority level has a set of queues at each apiserver; a PriorityLevelConfiguration is just an API object that configures a priority level and does not have queues.

Reviewer: How I'd help the reader here: have a fragment later that describes those API objects, and make “API objects” hyperlink to that fragment.

based on API objects that configure this feature.  This involves two
concepts: priority level and flow schema.

### Priority level

> Each apiserver has an aggregate concurrency limit and divides it ---

Reviewer: nit: Did you mean:
```suggestion
Each apiserver has an aggregate concurrency limit and divides it&mdash;
```
?

without examining the request load --- among several priority levels.
The priority levels operate independently of each other.  That is, the
apiserver classifies a given request to exactly one priority level

> and, for a given priority level, handles at most the corresponding

Reviewer 2: This section could be broken down and simplified.

> number of requests at any given time.  One special priority level ---

Reviewer: nit: Did you mean:
```suggestion
number of requests at any given time.  One special priority level,
```

> used for meta-requests and other top-priority traffic --- imposes no

Reviewer: How about:
```markdown
There is a special priority level, termed “exempt”, that does not impose any limit on concurrency.
```

Author: The original has a definition of "exempt" and says it applies to both priority levels and requests.

limit.  This priority level and the requests mapped to it are
described as "exempt", because of the lack of concurrency limitation.

For a non-exempt request, if it is not allowed to be handled
immediately then the apiserver will either reject the request or hold
it in a queue for a while.  Each non-exempt priority level has a
number of FIFO queues.  Whenever that priority level's number of
requests currently being handled drops below the concurrency limit,
the apiserver draws requests to begin handling from these queues using
a technique that is an appropriate generalization of
[fair queuing](https://en.wikipedia.org/wiki/Fair_queuing).

### Flow schema

To determine how to control a request, an apiserver first classifies
that request to a _flow_ according to a configured collection of _flow
schemas_.  The flow schemas are considered in order, according to each
flow schema's _matching precedence_ (which is a number).  A flow
schema has rules that describe which requests match that schema.

> Matching takes into account (a) the requested operation --- as

Reviewer: nit: Did you mean:
```suggestion
Matching takes into account (a) the requested operation&mdash;as
```
?

Author: I tried replacing ` --- ` with `&emdash;` but the rendering shows the `&emdash;`

Reviewer: `&emdash;` is the wrong name; try `&mdash;`

identified by the HTTP request verb and URI --- and (b) if the request
arrived through a secured port then the authenticated user attributes

> otherwise just the UserAgent.  Among the schemas that match the

Reviewer: How about starting a new sentence, eg:
```suggestion
For unauthenticated requests, the API server classifies the request based on its `User-Agent` header only.  Among the schemas that match the
```

Author: That would break up the logic, leaving an incorrect statement in part (b).

request, one with the logically highest matching precedence wins.

A flow schema also has a declaration of how to complete the _flow

> identifier_ for the request's flow.  A flow identifier is a pair of

Reviewer 2: Could you provide a code example (could be partial code sample) of a flow schema?

Author: Would pointing into the KEP be appropriate?

Reviewer 2: I would be in favor of a real example.
Which KEP are you referring to?

Reviewer: Opinions, based on what I've observed:

I think it's most common to put an example resource (eg in YAML) either in the Concept page or in a Task page. If the YAML is usable unmodified, and you feel like having a separate Task could make the Concept page more readable, by shortening it, then I favor that approach.

If the YAML is just an example that the reader can't use as-is, the Concept page is probably the best home for it (and it's OK to repeat that YAML in the Reference docs if appropriate).

strings: the name of its flow schema plus one more string called the
_flow distinguisher_.  A flow schema indicates whether the

> distinguisher is (1) the username or UserAgent of the user making the

Reviewer: If something other than a human user makes a request (eg a Pod), does that client's identity contribute to the _flow distinguisher_?

Author: I thought the language here answers that question.  It speaks of requests without limitation to those made by human using kubectl.  The sentence commented upon says that what contributes to the flow distinguisher is declared by the flow schema.

Reviewer: These docs usually use the term “user” to mean a human, so I'm keen to make sure the language helps a reader who (eg) hasn't yet come across ServiceAccount as a concept.

request, (2) the namespace (if any) of the object that the request
seeks to access, or (3) the empty string.

A flow schema also states the name of the priority level to use for
requests that match.

## Queuing

An apiserver first classifies a request to a flow, and thus a

> corresponding priority level, and if the priority level is not exempt

Reviewer: ```suggestion
corresponding priority level. If the priority level (of the flow) is not exempted from queuing,
```

Author: The intent here is to define "exempt" as a technical term and use it so.  There is a definition earlier, and the usage here is consistent with that definition.

Reviewer: To follow house style, make that first use of exempt be emphasized rather than in quotes.

> corresponding priority level, and if the priority level is not exempt

Reviewer 2: API server vs. apiserver

then the apiserver assigns the request to a queue using a technique
called "shuffle sharding".  This technique starts with hashing the
flow identifier, and then uses the hash value as a source of entropy
to pseudo-randomly pick a small subset of the queues, and finally
assigns the request to one of the shortest queues in that subset.  The
priority level declares the size of the subset.

After assigning a request to a priority level and a queue, the
apiserver immediately begins handling the request if that would not
violate the priority level's concurrency limit; otherwise the queue
holds the request until it can be executed or is evicted.

The apiserver applies two criteria to limit queue occupancy.  First,
each non-exempt priority level has a queue length limit.  The
apiserver will not add a request to a queue that is already at or
exceeding its length limit; rather, the request is immediately
rejected.  Second, each apiserver is configured with a limit on how
long a request can wait in a queue.  If that limit expires before the

> request can be handled then the request is rejected.  Rejection uses

Reviewer: I suggest:
_when rejecting requests, the API server responds using HTTP status 429, “Too Many Requests”_.

(rewording to use the active voice).

Author: Reworded to active voice.  I chose not to use two commas wrongly.  I often see HTTP status codes described with number and text run together, no comma between them.

HTTP status code 429 "Too Many Requests".

## Configuration

Configuration is mainly through API objects of kinds

> `PriorityLevelConfiguration` and `FlowSchema`, in the APIGroup

Reviewer: I suggest ”in the `flowcontrol.apiserver.k8s.io` API group”, which matches how existing pages phrase things.

Author: Thanks, done.

> `PriorityLevelConfiguration` and `FlowSchema`, in the APIGroup

Reviewer 2: Could you provide a configuration example?

`flowcontrol.apiserver.k8s.io`.

The apiserver's aggregate concurrency limit is the sum of the two
limits applied by the older simpler functionality: the limit on the

> number of mutating requests being handled at one time and the limit on

Reviewer: where does a cluster operator read or set those limits?

Author: With command line or config file parameters of the server program (i.e., `kube-apiserver` or an aggregated apiserver).  Is there a way to refer to the definitions of specific config items for `kube-apiserver`?

I expect there will be a REFERENCE section page about this too.  I am getting a little concerned about putting reference material in the CONCEPTS page.

the number of readonly requests being handled at a time.

An apiserver's limit on how long a request can wait in a queue is one

> quarter of the apiserer's handling timeout for non-long-running

Reviewer: nit: 
```suggestion
quarter of the apiserver's handling timeout for non-long-running
```

Reviewer: Aside: I think I would struggle to localize ”_handling timeout for non-long-running requests_” into another language.

Reviewer 2: 👍 . Lines 119 - 121 should be cleaned up and rephrased.

Author: I reworded.
